CRM/Contact/BAO/GroupContact.php

```diff
@@ -211,6 +213,12 @@ public static function removeContactsFromGroup(
           'tracking' => $tracking,
         ];
         CRM_Contact_BAO_SubscriptionHistory::create($historyParams);
+        // Removing a row from civicrm_group_contact for a smart group may mean a contact
+        // Is now back in a group based on criteria so we should re-load the cache.
+        $groupInformation = civicrm_api3('Group', 'getsingle', ['id' => $groupId]);
+        if (!empty($groupInformation['saved_search_id'])) {
+          CRM_Contact_BAO_GroupContactCache::load($groupId, TRUE);
+        }
```

Comment:
@seamuslee001 Instead of explicitly loading the cache here how about adding a new function to `CRM_Contact_BAO_GroupContactCache` that invalidates the cache for that group (meaning that it will be rebuilt next time it is needed):
```
public static function expireGroupContactCache($groupID) {
  $update = "
  UPDATE civicrm_group g
  SET    cache_date = null, refresh_date = null
  WHERE  id = %1 ";

  $params = [
    1 => [$groupID, 'Integer'],
  ];

  CRM_Core_DAO::executeQuery($query, $params);
}
```

That should cause the cache for that group to be rebuilt next time it is accessed.  And I think we could start using the function in a lot more places to stop the cache being rebuilt just because we've made a change, rather than because we actually need to use it!

Reply:
yes I agree if we invalidate rather than flush it the cache then it could be cleared by the relevant scheduled job with less uI impact